### DISCUSSION TITLE
Shutterbuttons on ESP32

### DISCUSSION THREAD

**Discussion by DaBaBW** [arendst/Tasmota#19165] [Q&A]
After my [last question](https://github.com/arendst/Tasmota/discussions/19133) was closed too early now this is still my problem:

I´m not able to configure more than 4 buttons. And configuration isn´t saved to file, too.
So I think the ESP32 functions of shutters are not active when compiled with `#define USE_SHUTTER` (or checkbox SHUTTERS AND BLINDS in online compiler) even in Tasmota 13.0.0.
Tested on Nodemcu 38pin ESP-32S and one with WROOM-32D with online compiler like the pictures below.

If I compile with `#define USE_SHUTTER_ESP32` configuration is saved to file and I´m able to configure all 8 buttons of a PCF8574 in Mode2 but only for Shutter1. Shutter 2 to 4 everytime reply `RSL: RESULT = {"ShutterButton4":"-"}`

![Screenshot 2023-07-21 184243](https://github.com/arendst/Tasmota/assets/36332774/05eb6da7-83d7-4ef0-81e2-de829eb745ae)
![Screenshot 2023-07-21 184428](https://github.com/arendst/Tasmota/assets/36332774/a9e9185f-7c60-4e0c-9761-2bd48b1bbfe5)
![Screenshot 2023-07-21 184851](https://github.com/arendst/Tasmota/assets/36332774/08332d9b-8489-44aa-83a6-219bbdc09480)
![ESP32 Konfig 4-fach Shutter](https://github.com/arendst/Tasmota/assets/36332774/5c0608c5-ac5e-4145-ae4e-20d5b5249bde)

**Comment by Jason2866** [comment]
The last esp32 shutter changes are in development branch. You build is from release branch.

**Comment by DaBaBW** [comment]
Ah ok. But I can´t choose development in online compiler. So I have to do it another way?

**Reply by Jason2866** [reply]
As explained in the Tasmota Documentation another way without a local PC is with Gitpod.

**Reply by barbudor** [reply]
As explained in the TasmoCompiler documentation, to build from Tasmota Development you need to choose TasmoCompiler Development

**Comment by DaBaBW** [comment]
Thank you very much, now I was able to compile a development bin. But this did not change my problem.
No shutter buttons for more than Shutter1...

**Comment by sternegugger** [comment]
I have a similar problem: I can set ShutterButton1, but not higher.
Firmware from latest Development Branch (20. August 2023 tasmota32-DE) with ''#define USE_MCP23XXX_DRV''.
I am using an ESP32 with 3 MCP23017 to control 24 relays and 24 switches for up to 12 shutters. 
Configuration:
<code>
rule1 on file#mcp23x.dat do {"NAME":"MCP23017 shutter","BASE":0,"GPIO":[160,161,162,163,164,165,166,167,256,257,258,259,260,261,262,263, 168,169,170,171,172,173,174,175,264,265,266,267,268,269,270,271, 176,177,178,179,180,181,182,183,272,273,274,275,276,277,278,279]} endon

backlog
SetOption80 1;
PowerOnState 0;
ShutterMode 2;
Shutterrelay1 1;
Shutterrelay2 3;
Shutterrelay3 5;
Shutterrelay4 7;
...
backlog
SetOption32 10;
SetOption1 1;
SetOption73 1;
ButtonTopic 0;
setOption73 1;

backlog 
SetOption114 1;
SwitchMode1 5;
SwitchMode2 5;
SwitchMode3 2;
SwitchMode4 2;
...
backlog 
ShutterButton1 1 up 1;
ShutterButton1 2 down 1;
ShutterButton2 3 up 1;
ShutterButton2 4 down 1;
...
</code>
ShutterButton1 can be configurated normally. But ShutterButton2 does not accept the configuration (see the last 5 lines in Log):
<code>
17:37:48.326 CMD: backlog  ShutterButton1 1 up 1; ShutterButton1 2 down 1; ShutterButton2 3 up 1;
17:37:48.328 SRC: WebConsole from 192.168.2.141
17:37:48.330 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 70, Pld -99, Data 'ShutterButton1 1 up 1; ShutterButton1 2 down 1; ShutterButton2 3 up 1;'
17:37:48.359 SRC: Backlog
17:37:48.361 CMD: Grp 0, Cmd 'SHUTTERBUTTON', Idx 1, Len 6, Pld 1, Data '1 up 1'
17:37:48.364 SHT: FINAL: pos:0, inc:0, string:up
17:37:48.365 SHT: FINAL: pos:1, inc:0, string:1
17:37:48.366 SHT: CASE4: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.367 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.369 SHT: Merging: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.371 SHT: Merging: pos:75, tilt:-128, incpos:0, inctilt:0
17:37:48.373 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.376 SHT: Merging: pos:0, tilt:-128, incpos:0, inctilt:0
17:37:48.378 SHT: Merging: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.380 SHT: Merging: pos:25, tilt:-128, incpos:0, inctilt:0
17:37:48.382 SHT: Merging: pos:0, tilt:-128, incpos:0, inctilt:0
17:37:48.384 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.386 SHT: Merging: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.388 SHT: Merging: pos:75, tilt:-128, incpos:0, inctilt:0
17:37:48.390 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.397 MQT: Rolladen/stat/RESULT = {"ShutterButton1":"1 100 1 75 100 - - - -|2 0 1 25 0 - - - -|3 100 1 75 100 - - - -"}
17:37:48.611 SRC: Backlog
17:37:48.614 CMD: Grp 0, Cmd 'SHUTTERBUTTON', Idx 1, Len 8, Pld 2, Data '2 down 1'
17:37:48.617 SHT: FINAL: pos:0, inc:0, string:down
17:37:48.619 SHT: FINAL: pos:1, inc:0, string:1
17:37:48.621 SHT: CASE4: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.622 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.623 SHT: Merging: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.626 SHT: Merging: pos:75, tilt:-128, incpos:0, inctilt:0
17:37:48.627 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.629 SHT: Merging: pos:0, tilt:-128, incpos:0, inctilt:0
17:37:48.632 SHT: Merging: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.633 SHT: Merging: pos:25, tilt:-128, incpos:0, inctilt:0
17:37:48.635 SHT: Merging: pos:0, tilt:-128, incpos:0, inctilt:0
17:37:48.637 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.639 SHT: Merging: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.641 SHT: Merging: pos:75, tilt:-128, incpos:0, inctilt:0
17:37:48.643 SHT: Merging: pos:100, tilt:-128, incpos:0, inctilt:0
17:37:48.654 MQT: Rolladen/stat/RESULT = {"ShutterButton1":"1 100 1 75 100 - - - -|2 0 1 25 0 - - - -|3 100 1 75 100 - - - -"}
17:37:48.816 SRC: Backlog
17:37:48.818 CMD: Grp 0, Cmd 'SHUTTERBUTTON', Idx 2, Len 6, Pld 3, Data '3 up 1'
17:37:48.821 SHT: FINAL: pos:0, inc:0, string:up
17:37:48.824 SHT: FINAL: pos:1, inc:0, string:1
17:37:48.825 SHT: CASE4: pos:1, tilt:-128, incpos:0, inctilt:0
17:37:48.835 MQT: Rolladen/stat/RESULT = {"ShutterButton2":"-"}
</code>

Did I miss something in the configuration?
Should I create an new issue?
@stefanbode  It may interact with your Issue #18295

**Reply by stefanbode** [reply]
Thanks for bringing it up. Agree that it should work. Let me test why it reset to -

**Reply by sternegugger** [reply]
@stefanbode the fix #19362 (shutterinvert ESP32) seems to fix the shutterButton problem too.
With development version from 2023-08-22 15:00 I can now set all shutterButtons including the higher numbers.
<code>
09:03:38.719 MQT: Rolladen/stat/RESULT = {"ShutterButton11":"21 100 1 75 100 - - - -|22 0 1 25 0 - - - -"}
</code>
